src/realm/array_decimal128.cpp

```diff
@@ -81,11 +174,45 @@ size_t ArrayDecimal128::find_first(Decimal128 value, size_t start, size_t end) c
         end = sz;
     REALM_ASSERT(start <= sz && end <= sz && start <= end);
 
-    auto values = reinterpret_cast<Decimal128*>(this->m_data);
-    for (size_t i = start; i < end; i++) {
-        if (values[i] == value)
-            return i;
+    auto width = min_width(value);
+    switch (m_width) {
+        case 0:
+            if (value.is_null()) {
+                return 0;
+            }
+            break;
+        case 4:
+            if (width <= 4) {
+                // Worth optimizing here
+                auto optval32 = value.to_bid32();
+                REALM_ASSERT(optval32);
+                auto val32 = *optval32;
+                auto values = reinterpret_cast<Decimal128::Bid32*>(this->m_data);
+                for (size_t i = start; i < end; i++) {
+                    if (values[i] == val32)
+                        return i;
+                }
+            }
+            break;
+        case 8:
+            if (width <= 8) {
+                auto values = reinterpret_cast<Decimal128::Bid64*>(this->m_data);
+                for (size_t i = start; i < end; i++) {
+                    if (Decimal128(values[i]) == value)
```

Comment:
Can this be done by comparing Bid64 values directly, instead of converting up to Decimal128?

Reply:
That could be done, but then I would have to implement the comparison function like in the 32 bit case. But I don't think the gain would be that big as it is relatively quick to upgrade. Can be done, if you insist.

Reply:
Not a blocker since it can be optimized later without changing the file format if it actually does show up in performance profiling.